core/src/main/java/smile/classification/RBFNetwork.java

```diff
@@ -156,7 +156,7 @@ public Trainer(Metric<T> distance) {
          * @param rbf the radial basis function.
          * @param m the number of basis functions.
          */
-        public Trainer setRBF(RadialBasisFunction rbf, int m) {
+        public Trainer<T> setRBF(RadialBasisFunction rbf, int m) {
```

Comment:
At first, no this isn't a Java 8 feature, this is in place since Generics are, I think this is since 1.5.

What happens there? Well Trainer itself is a class with a generic type parameter T (c.f. line 127 in the code, later on used for the distance field (`Metric<T>`)) and as such it should not refer to the "raw type" `Trainer` (without generic parameter) as you loose compile-time safety about the types used there.
Think of e.g. Lists, when you have a list of strings you write `List<String>` and get only offered methods to add strings to this list, and of course you can also only retrieve strings from that List (no ints, doubles, other types). But an even better example is the method `subList` which returns a List of Strings again, if they did omit the generic type parameter here, only a list would be returned where you do not know of which type the elements are, you would get only Objects instead everywhere. (This is actually what happens at runtime due to type erasure, but this is another topic ;) )

So in your specific case, the reason is that you wouldn't know which type is inside the trainer instance afterwards besides by using casts, which are not necessary if generics are used consistently.